### DISCUSSION TITLE
Easiest way to edit & insert larger code sections in to a program?

### DISCUSSION THREAD

**Discussion by nickhfoley** [visose/Robots#259] [Q&A] [answered]
I'd like to start my programs with some longer sections of code that trigger multiple teach pendant interactions with conditionals, move to positions, set DOs, etc. and it will be much easier to hand code this in RAPID vs. build them line by line as grasshopper nodes. 

Is there an easy way I can work in a text editor (within grasshopper?) to feed a code block into the Init Commands part of the Program element, for example? I'm versed in using the Script elements like ghPython in a sub-window to write code and then pull it into a larger Grasshopper program and would like to do something similar with RAPID into Robots nodes. 

I suppose this could even just be a text editor within Grasshopper that then passed the contents along with the correct syntax, but I'm not sure if that exists. 

How would you propose to accomplish this?

**Comment by visose** [comment]
The best editor for RAPID is likely ABB's Robot Studio. There are also some extensions for Visual Studio Code that have syntax highlighting for RAPID. In either case the code will be saved to a text file which you can load to Grasshopper using the "Read File" component.

**Reply by nickhfoley** [reply]
Noted, and this is helpful - I especially did not know of the Visual Studio extension. However, I'm still hoping for a way to do this coding directly in Grasshopper, so that I can automatically pass geometry-specific details into the routine (starting points based on loaded parts). I'm also bringing in Grasshopper + Robots with an effort to keep everything within a single tool, vs the balkanized mess of our prior processing flow.

I suspect I can find a solution using one of the scripting components, either outputting RAPID directly or automatically making edits to a text file that is read back in.

**Comment by visose** [comment, marked as answer]
To keep longer RAPID/KRL/URScript blocks editable, use a text file plus Grasshopper's Read File component, a Panel, or a scripting component that outputs a string. Wrap raw robot-language snippets in Custom Command when they should run as init/target commands, or use Custom Code when you want to replace the whole generated program text. The Create Program `Init Commands` input expects Command objects, not raw text directly.
